### DISCUSSION TITLE
How to get the lsp to find types in other local files?

### DISCUSSION THREAD

**Discussion by realh** [vknabel/vscode-swift-development-environment#90] [Q&A]
I'm very new to Swift and Apple development, coming from a Linux background. I'm trying to write a simple game for a Mac, on a Mac, but I'd rather use VS Code than Xcode so I don't have to waste hours resigning the whole IDE after every update to make the vim plugin work.

So far I've only written 2 very short swift files for my project, but this seems to already be too complicated for sourcekit-lsp to deal with. I have a protocol in one file (TextureManager) and a class that implements it (AppleTextureManager) in another file. This causes an error in VS Code: `cannot find type 'TextureManager' in scope` (followed by `sourcekitd` in a paler font). The two files are in separate xcode groups/folders, but even when I tried moving one of them so they're both in the same folder the error won't go away. 

What do I have to do to make this plugin work with multiple source files?

**Comment by vknabel** [comment]
Hi @realh! Thanks for reaching out!

`sourcekit-lsp` requires a Swift Package Manager project to work correctly. [This blog post on theswiftdev.com](https://theswiftdev.com/how-to-build-macos-apps-using-only-the-swift-package-manager/) seems to fit your case perfectly! Migrating your project to Swift Package Manager (SwiftPM or SPM) is the easiest way and **I really recommend using it**. Plus: you can use SPM to write command line tools and servers, that also support Linux.

---

In case you need some additional resources, I wrote two blog posts, which might help you later: [Debugging Swift in VS Code](https://www.vknabel.com/pages/Debugging-Swift-in-VS-Code/) and [Swift VS Code and you](https://www.vknabel.com/pages/Swift-VS-Code-and-you/). But you can take one step at a time.

---

If you explicitly don't want to use SPM, you have to use the more flexible `sourcekite`. It is much harder to configure and I don't recommend beginners to use it.

Here you'd have multiple options:
- it supports SPM projects, too
- By default it puts all Swift files into one target (if they have none)
- The `swift.targets` setting allows defining `compilerArguments` for specific groups of swift files. Pretty low level.

---

Hopefully I could help you, get a good start!

**Reply by realh** [reply]
Yes, this is very helpful, thanks. I also got a reply in the Swift forums, but they didn't go the extra mile and show how I can build a Mac app with SPM. However, I've decided to just put up with resigning Xcode for now, because the extra hand-holding should be helpful for a newbie to Apple like me. My aim is make a portable game, going as native as possible/practical on each platform; so SpriteKit for Apple, because it's the easiest way to use Metal for 2D, and I can easily write a fake version for other platforms, implemented in OpenGL (ES). So once I've got the basics up and running on the Mac I should easily be able to do the bulk of the development on Linux in case I get fed up with Xcode.

**Reply by vknabel** [reply]
Of course, Xcode helps a lot. Especially the visual editors are great for beginners to discover what’s possible!
